### DISCUSSION TITLE
For Ver 3.4.1 GLIBC_2.33, GLIBC_2.34 unresolved on Linux Mint

### DISCUSSION THREAD

**Discussion by topchyan** [pawelsalawa/sqlitestudio#4631] [General]
### Details
Error messages I get after installing 3.4.1
./sqlitestudio: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by ./sqlitestudio)
./sqlitestudio: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /opt/SQLiteStudio/lib/libsqlite3.so)
./sqlitestudio: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /opt/SQLiteStudio/lib/libsqlite3.so)

### Steps to reproduce
sudo tar xf ~/Downloads/sqlitestudio-3.3.3.tar.xz -C /opt

### Operating system
[code]
System:    Kernel: 5.4.0-135-generic x86_64 bits: 64 compiler: gcc v: 9.4.0 Desktop: Xfce 4.16.0 
           tk: Gtk 3.24.20 wm: xfwm4 dm: LightDM Distro: Linux Mint 20.3 Una 
           base: Ubuntu 20.04 focal 
Machine:   Type: Convertible System: HP product: HP Pavilion x360 Convertible 11m-ap0xxx 
           v: Type1ProductConfigId serial: <filter> Chassis: type: 31 serial: <filter> 
           Mobo: HP model: 85CA v: 26.18 serial: <filter> UEFI: Insyde v: F.01 date: 01/16/2019 
Battery:   ID-1: BAT0 charge: 33.2 Wh condition: 33.2/33.2 Wh (100%) volts: 8.5/7.6 
           model: 333-1C-3D-A PO02037XL serial: <filter> status: Full 
           Device-1: hid-0018:04F3:2420.0002-battery model: ELAN0732:00 04F3:2420 serial: N/A 
           charge: N/A status: N/A 
CPU:       Topology: Quad Core model: Intel Pentium Silver N5000 bits: 64 type: MCP 
           arch: Goldmont Plus rev: 1 L2 cache: 4096 KiB 
           flags: lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx bogomips: 8755 
           Speed: 2587 MHz min/max: 800/2700 MHz Core speeds (MHz): 1: 2586 2: 2680 3: 2587 
           4: 2587 
Graphics:  Device-1: Intel UHD Graphics 605 vendor: Hewlett-Packard driver: i915 v: kernel 
           bus ID: 00:02.0 chip ID: 8086:3184 
           Display: x11 server: X.Org 1.20.13 driver: modesetting unloaded: fbdev,vesa 
           resolution: 1366x768~60Hz 
           OpenGL: renderer: Mesa Intel UHD Graphics 605 (GLK 3) v: 4.6 Mesa 21.2.6 
           direct render: Yes 
Audio:     Device-1: Intel vendor: Hewlett-Packard driver: snd_hda_intel v: kernel bus ID: 00:0e.0 
           chip ID: 8086:3198 
           Sound Server: ALSA v: k5.4.0-135-generic 
Network:   Device-1: Realtek RTL8821CE 802.11ac PCIe Wireless Network Adapter 
           vendor: Hewlett-Packard driver: rtl8821ce v: v5.5.2.1_35598.20191029 port: 1000 
           bus ID: 02:00.0 chip ID: 10ec:c821 
           IF: wlo1 state: up mac: <filter> 
Drives:    Local Storage: total: 357.54 GiB used: 214.47 GiB (60.0%) 
           ID-1: /dev/sda vendor: LITE-ON model: CV8-8E128-HP size: 119.24 GiB speed: 6.0 Gb/s 
           serial: <filter> 
           ID-2: /dev/sdb type: USB model: Multiple Card Reader size: 238.30 GiB serial: <filter> 
Partition: ID-1: / size: 116.32 GiB used: 46.33 GiB (39.8%) fs: ext4 dev: /dev/sda5 
USB:       Hub: 1-0:1 info: Full speed (or root) Hub ports: 9 rev: 2.0 chip ID: 1d6b:0002 
           Device-1: 1-4:2 info: Quanta HP Wide Vision HD Camera type: Video driver: uvcvideo 
           rev: 2.0 chip ID: 0408:5300 
           Device-2: 1-5:3 info: Alcor Micro Multi Flash Reader type: Mass Storage 
           driver: usb-storage rev: 2.0 chip ID: 058f:6366 
           Device-3: 1-6:4 info: Realtek Bluetooth Radio type: Bluetooth driver: btusb rev: 1.1 
           chip ID: 0bda:b00a 
           Hub: 2-0:1 info: Full speed (or root) Hub ports: 7 rev: 3.0 chip ID: 1d6b:0003 
Sensors:   System Temperatures: cpu: 56.0 C mobo: 33.0 C 
           Fan Speeds (RPM): N/A 
Repos:     No active apt repos in: /etc/apt/sources.list 
           Active apt repos in: /etc/apt/sources.list.d/official-package-repositories.list 
           1: deb http: //packages.linuxmint.com una main upstream import backport #id:linuxmint_main
           2: deb http: //archive.ubuntu.com/ubuntu focal main restricted universe multiverse
           3: deb http: //archive.ubuntu.com/ubuntu focal-updates main restricted universe multiverse
           4: deb http: //archive.ubuntu.com/ubuntu focal-backports main restricted universe multiverse
           5: deb http: //security.ubuntu.com/ubuntu/ focal-security main restricted universe multiverse
           6: deb http: //archive.canonical.com/ubuntu/ focal partner
           Active apt repos in: /etc/apt/sources.list.d/vscode.list 
           1: deb [arch=amd64,arm64,armhf] http: //packages.microsoft.com/repos/code stable main
           Active apt repos in: /etc/apt/sources.list.d/yannubuntu-boot-repair-focal.list 
           1: deb http: //ppa.launchpad.net/yannubuntu/boot-repair/ubuntu focal main
Info:      Processes: 212 Uptime: 13h 03m Memory: 3.41 GiB used: 1.68 GiB (49.1%) Init: systemd 
           v: 245 runlevel: 5 Compilers: gcc: 9.4.0 alt: 9 Client: Unknown python3.8 client 
           inxi: 3.0.38 
[/code]

### SQLiteStudio version
3.4.1

**Comment by pawelsalawa** [comment]
> base: Ubuntu 20.04 focal

This is the problem. Binaries of 3.4.x are built on Ubuntu 22.04 and won't work on Ubuntu 20.04 (which your Mint is based on). You need to upgrade your Mint in order to run 3.4.1 binaries. Otherwise binaries dedicated for Ubuntu 20.04 would need to be compiled.

**Reply by topchyan** [reply]
Pawel, thanks for your reply. Unfortunately many other important programs they just break upon upgrade to 22.04, thus I remain on the stable 20.04 version for a foreseeable future. Or perhaps if I change the machine I'm on I'll try the new one.

**Reply by orangelynx** [reply]
Is there any particular reason why you cannot compile the binaries to be backwards compatible? GLIBC for sure is very backwards compatible. All kinds of software packages are compiled on the latest LTS or even totally unrelated distributions yet work fine on older versions of the same or other distributions. I think you are unnecessarily diminishing your user base by this restriction.

**Reply by pawelsalawa** [reply]
It is backward compatible, but not forward compatible.

Applications compiled on Ubuntu 20.04 generally work fine on Ubuntu 22.04, but not the other way around. That's why you're experiencing problems on 20.04, as SQLiteStudio is compiled on 22.04.

Why is it compiled on 22.04? Frankly speaking I don't remember exact reason. As far as I try to remember, there was another issue caused by 20.04, that was fixed on 22.04, but I may remember it wrong.

Nevertheless, we're in 2023. One may as why don't I compile on Ubuntu 14.04, so it works on all subsequent Ubuntu releases.

**Reply by orangelynx** [reply]
Well, admittedly, it seems to be more complicated than I would have though to compile against an older version of GLIBC that just support what's actually being used (see e.g. https://github.com/wheybags/glibc_version_header), and the common recommendation is indeed to build on the oldest supported system (ludicrous in my opinion). However, if you don't need features / bugfixes included in the libraries shipping with 22.04 LTS, I'd encourage you to use a container building environment or something like AppImage to ensure that 20.04 LTS users can still use this great piece of software in its latest version. I think it's reasonable to support Ubuntu versions which are still officially supported by canonical.

Apparently, often just few symbols need to be changed: https://stackoverflow.com/questions/4032373/linking-against-an-old-version-of-libc-to-provide-greater-application-coverage

**Reply by pawelsalawa** [reply]
Generally it's not difficult to build on 20.04 using GitHub Actions (which I do use here). First I need to look up what was the reason to use 22.04.

**Reply by pawelsalawa** [reply]
It seems that everything works fine. At least at first glance. You can get Ubuntu 20.04 packages from Action artifacts: https://github.com/pawelsalawa/sqlitestudio/actions/runs/3952996538

SQLiteStudio 3.4.3 will be built on Ubuntu 20.04 (if nothing wrong turns out from this build here).

**Reply by damufo** [reply]
Very thanks.
Work fine on xubuntu 20.04.

**Reply by orangelynx** [reply]
Looks good! On Ubuntu 20.04 LTS

**Reply by pawelsalawa** [reply]
Great. I hope to release 3.4.3 soon (only one bug remaining to fix at the moment). It will be built on Ubuntu 20.04.

**Comment by damufo** [comment]
Same problem for Ubuntu 20.04 LTS.

**Reply by pawelsalawa** [reply]
As I already mentioned in the answer above - binaries are built on Ubuntu 22.04 and won't work on Ubuntu 20.04 anymore.
